### PR TITLE
feat(python): support sequence packing in streaming dataset

### PR DESCRIPTION
**Pull request by AyushExel**
## How packing works

  Consider four tokenized documents:

  
  [1]
  [2]
  [10, 11, 12, 13, 14, 15, 16, 17]
  [20]

  With:
```
  StreamingDataset(
      table,
      shuffle=False,
      columns=["tokens"],
      num_splits=2,
      pack_sequences=5,
      eos_id=9,
      pad_id=0,
      blocks_per_epoch=6,
  )
```
  the documents are assigned to two fixed logical splits. Each split maintains an independent token buffer, appends eos_id after every document, and emits blocks of five tokens.

  Because blocks_per_epoch=6, each split emits exactly three blocks:

  Cycl/e 1:
    Split 0: [1, 9, 2, 9, 0] # 9 is eos, 0 is padding
    Split 1: [10, 11, 12, 13, 14]

  Cycle 2:
    Split 0: [0, 0, 0, 0, 0]
    Split 1: [15, 16, 17, 9, 20]

  Cycle 3:
    Split 0: [0, 0, 0, 0, 0]
    Split 1: [9, 0, 0, 0, 0]

  If a split runs out of tokens early, it emits padded blocks through the fixed budget. This prevents one rank from finishing before another.

  Logical splits are independent of rank and worker ownership. A checkpoint records each split’s consumed-document count, emitted-block count, remaining tokens, and document boundaries. Merging
  those per-split states allows the same packed stream to resume after the topology changes.

  doc_ids identifies document segments, including continuations across block boundaries. It is not a padding mask: padding retains the preceding document ID, so callers must mask padding using a
  reserved pad_id.

  blocks_per_epoch="auto" is also available. It estimates the budget from a deterministic bounded sample and warns that the result is approximate.

WIP pre-training tests:
```
  ┌────────────────────────────────────┬────────────────────────┬─────────────────────────────────────┐
  │                                    │       GPT-2 124M       │          GPT-2 medium 354M          │
  ├────────────────────────────────────┼────────────────────────┼─────────────────────────────────────┤
  │ Corpus                             │ 2.4M docs / 12GB table │ 9.67M docs / 45GB table             │
  │ Tokens (Chinchilla)                │ 2.43B                  │ 7.0B                                │
  ├────────────────────────────────────┼────────────────────────┼─────────────────────────────────────┤
  │ Data prep (ingest→curate→tokenize) │ ~12 min                │ ~51 min                             │
  ├────────────────────────────────────┼────────────────────────┼─────────────────────────────────────┤
  │ Training wall time                 │ ~50 min                │ 3h 06m                              │
  ├────────────────────────────────────┼────────────────────────┼─────────────────────────────────────┤
  │ Throughput / MFU                   │ 1.60M tok/s / 35%      │ 684k tok/s / 42.0%, │
  ├────────────────────────────────────┼────────────────────────┼─────────────────────────────────────┤
  │ Final val loss                     │ 3.230                  │ 2.840                               │
  └────────────────────────────────────┴────────────────────────┴─────────────────────────────────────┘
```

### CODE REVIEW

**Comment by AyushExel** [comment]
@claude review

File: python/python/lancedb/streaming.py [hunk 1]
```diff
@@ -337,6 +398,13 @@ def __iter__(self) -> Iterator[dict[str, Any]]:
         my_splits = self._resolve_my_splits()
         if not my_splits:
             return
+        if self._pack_sequences is not None and len(my_splits) < self._num_splits:
```
[the remark this answers is not in the document]
Comment (AyushExel) [reply python/python/lancedb/streaming.py]: fixed

File: python/python/lancedb/streaming.py [hunk 2]
```diff
@@ -291,13 +393,83 @@ def __init__(
...
+        rng = random.Random(self._shuffle_seed)
+
+        warnings.warn(
+            "blocks_per_epoch='auto' uses an approximate token-count sample; "
+            "pass an explicit value for exact epoch sizing",
+            stacklevel=3,
+        )
+
+        # TODO: Replace this fallback with Lance's dedicated exact token-count
+        # estimation API once it is available.
```
Comment (AyushExel) [inline python/python/lancedb/streaming.py]: This can be removed once we expose counting tokens for list<> tyeps

File: python/python/lancedb/streaming.py [hunk 3]
```diff
@@ -529,6 +614,54 @@ def _ensure_cooked(i: int) -> None:
...
+            doc_ids.cumsum_(dim=0)  # cumulative sum marks document boundaries
+            block = {
+                "input_ids": torch.tensor(tokens[:pack_len], dtype=torch.int64),
+                "doc_ids": doc_ids,
+            }
+            del tokens[:pack_len]
+            # Shift start boundaries for the next call.
+            buf["starts"] = [s - pack_len for s in starts if s >= pack_len]
+            return block
+
```
[the remark this answers is not in the document]
Comment (AyushExel) [reply python/python/lancedb/streaming.py:830]: Addressed the dead pack-mode position state and replaced the per-cycle deepcopy with an explicit structural copy. I kept the three Arrow list checks inline because `_is_list_like` is a private helper owned by `table.py`; importing it here would add cross-module coupling for a small, self-contained validation.

File: python/python/lancedb/streaming.py [hunk 4]
```diff
@@ -529,6 +614,54 @@ def _ensure_cooked(i: int) -> None:
...
+                if not cooked[i]:
+                    return
+                buf["starts"].append(len(buf["tokens"]))
+                pos, tokens = cooked[i].popleft()
+                buf["tokens"].extend(tokens)
+                buf["tokens"].append(eos_id)
+                pack_consumed[my_splits[i]] += 1
+                local_consumed[i] += 1
+                pos_consumed[i] = pos + 1
+                _advance(i)
```
[the remark this answers is not in the document]
Comment (AyushExel) [reply python/python/lancedb/streaming.py:811]: Addressed: the packing transform now rejects null token lists and null token values with a descriptive `ValueError` before they reach `_fill_block`.

File: python/python/lancedb/streaming.py [hunk 5]
```diff
@@ -216,6 +246,30 @@ def __init__(
...
+            field = table.schema.field(columns[0])
+            if not (
+                pa.types.is_list(field.type)
+                or pa.types.is_large_list(field.type)
+                or pa.types.is_fixed_size_list(field.type)
+            ):
+                raise ValueError(
+                    f"pack_sequences requires a list-typed token column; "
+                    f"{columns[0]} has type {field.type}"
+                )
```
[the remark this answers is not in the document]
Comment (AyushExel) [reply python/python/lancedb/streaming.py:323]: Addressed: construction now requires the list value type to be an Arrow integer type and reports the invalid value type directly.

**Comment by AyushExel** [comment]
@claude review

File: python/python/lancedb/streaming.py [hunk 6]
```diff
@@ -763,21 +1038,31 @@ def __setstate__(self, state):
...
+        if self._pack_sequences is not None:
+            return {
+                "shuffle_seed": self._shuffle_seed,
+                "num_splits": self._num_splits,
+                "epoch": self._epoch,
+                "pack_sequences": self._pack_sequences,
+                "eos_id": self._eos_id,
+                "pad_id": self._pad_id,
+                "blocks_per_epoch": self._blocks_per_epoch,
+                "samples_consumed_per_split": list(self._pack_consumed),
```
[the remark this answers is not in the document]
Comment (AyushExel) [reply python/python/lancedb/streaming.py:1107]: Confirmed this is an existing `StreamingDataset` multiprocessing limitation, not specific to sequence packing. The same parent-state reproduction fails in ordinary row mode. Its tracked here https://github.com/lancedb/lancedb/issues/3967 and it should not be related to this PR.

**Review by Xuanwo — APPROVED** [review]
